src/java.desktop/share/classes/javax/swing/JPasswordField.java

```diff
@@ -285,21 +285,27 @@ public String getText(int offs, int len) throws BadLocationException {
     public void setText(String t) {
         // overwrite the old data first
         Document doc = getDocument();
-        int nleft = doc.getLength();
-        Segment text = new Segment();
-        // we would like to get direct data array access, not a copy of it
-        text.setPartialReturn(true);
-        int offs = 0;
-        try {
-            while (nleft > 0) {
-                doc.getText(offs, nleft, text);
-                Arrays.fill(text.array, text.offset,
-                            text.count + text.offset, '\u0000');
-                nleft -= text.count;
-                offs += text.count;
+        DocumentFilter filter = null;
+        if (doc instanceof AbstractDocument) {
+            filter = ((AbstractDocument) doc).getDocumentFilter();
```

Comment:
The change is not clean because I have to remove pattern matching in this `instanceof `block.